src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.ChecksumToSolutionCache.SolutionAndInFlightCount.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal sealed partial class RemoteWorkspace
+    {
+        private sealed partial class ChecksumToSolutionCache
+        {
+            /// <summary>
+            /// Wrapper around asynchronously produced solution.  The computation for producing the solution will be
+            /// canceled when the number of in-flight operations using it goes down to 0.
+            /// </summary>
```

Comment:
i moved away from ReferenceCountedDisposable.  It was getting very hard to track what was going on, and the indirections it forced added a lot of complexity around locking and updating internal state here.  By using a private type for the cache, and the item being cached, we can safely and uniformly lock around everything, which enables us to perform all operations, synchronously, atomically and sanely.